pceplib/pcep_msg_messages_encoding.c

```diff
@@ -0,0 +1,304 @@
+    case PCEP_TYPE_OPEN:
+    case PCEP_TYPE_KEEPALIVE:
+    case PCEP_TYPE_PCREQ:
+    case PCEP_TYPE_PCREP:
+    case PCEP_TYPE_PCNOTF:
+    case PCEP_TYPE_ERROR:
+    case PCEP_TYPE_CLOSE:
+    case PCEP_TYPE_REPORT:
+    case PCEP_TYPE_UPDATE:
+    case PCEP_TYPE_INITIATE:
+        break;
+    default:
+        pcep_log(LOG_INFO, "Invalid PCEP message header type [%d]", msg_type);
+        return false;
+        break;
+    }
+
+    return true;
+}
+
+/* Internal util function */
+static void pcep_decode_msg_header(const uint8_t *msg_buf, uint8_t *msg_version, uint8_t *msg_flags, uint8_t *msg_type, uint16_t *msg_length)
+{
+    *msg_version = (msg_buf[0] >> 5) & 0x07;
+    *msg_flags = (msg_buf[0] & 0x1f);
+    *msg_type = msg_buf[1];
+    uint16_t *uint16_ptr = (uint16_t *) (msg_buf + 2);
+    *msg_length = ntohs(*uint16_ptr);
+}
+
+/* Decode the message header and return the message length */
+int16_t pcep_decode_validate_msg_header(const uint8_t *msg_buf)
+{
+    uint8_t msg_version;
+    uint8_t msg_flags;
+    uint8_t msg_type;
+    uint16_t msg_length;
+
+    pcep_decode_msg_header(msg_buf, &msg_version, &msg_flags, &msg_type, &msg_length);
+
+    return((validate_msg_header(msg_version, msg_flags, msg_type, msg_length) == false) ? -1 : (int16_t) msg_length);
+}
+
+bool validate_message_objects(struct pcep_message *msg)
+{
+    if (msg->msg_header->type >= PCEP_TYPE_UNKOWN_MSG)
+    {
+        pcep_log(LOG_INFO, "Rejecting received message: Unknown message type [%d]",
+                msg->msg_header->type);
+        return false;
+    }
+
+    const enum pcep_object_classes *object_classes = MANDATORY_MESSAGE_OBJECT_CLASSES[msg->msg_header->type];
+    double_linked_list_node *node;
+    int index;
+    for (node = (msg->obj_list == NULL ? NULL : msg->obj_list->head), index = 0;
+         index < NUM_CHECKED_OBJECTS;
+         index++, (node = (node==NULL ? NULL : node->next_node)))
+    {
+        struct pcep_object_header *obj = ((node == NULL) ? NULL : (struct pcep_object_header*) node->data);
+
+	if ((int)object_classes[index] == NO_OBJECT) {
+		if (node != NULL) {
+			pcep_log(
+				LOG_INFO,
+				"Rejecting received message: Unexpected object [%d] present",
+				obj->object_class);
+			return false;
+		}
+	} else if (object_classes[index] != ANY_OBJECT) {
+		if (node == NULL) {
+			pcep_log(
+				LOG_INFO,
+				"Rejecting received message: Expecting object in position [%d], but none received",
+				index);
+			return false;
+		} else if (object_classes[index] != obj->object_class) {
+			pcep_log(
+				LOG_INFO,
+				"Rejecting received message: Unexpected Object Class received [%d]",
+				object_classes[index]);
+			return false;
+		}
+	}
+    }
+
+    return true;
+}
+
+struct pcep_message *pcep_decode_message(const uint8_t *msg_buf)
+{
+    uint8_t msg_version;
+    uint8_t msg_flags;
+    uint8_t msg_type;
+    uint16_t msg_length;
+
+    pcep_decode_msg_header(msg_buf, &msg_version, &msg_flags, &msg_type, &msg_length);
```

Comment:
no call to `pcep_decode_validate_msg_header` first?

Reply:
This call to pcep_decode_msg_header is inside of pce_decode_validate_msg_header.